dv/uvm/core_ibex/sim.py

```diff
@@ -1,301 +1,512 @@
+    '''Substitute options and environment variables in cmd
+
+    enable_dict should be a dict mapping names to bools. For each key, N, in
+    enable_dict, if enable_dict[N] is False, then all occurrences of <N> in cmd
+    will be replaced with ''. If enable_dict[N] is True, all occurrences of <N>
+    in cmd will be replaced with opts_dict[N].
+
+    If N is not a key in opts_dict, this is no problem unless cmd contains
+    <N>, in which case we throw a RuntimeError.
+
+    Finally, the environment variables are substituted as described in
+    subst_env_vars and any newlines are stripped out.
+
+    '''
+    for name, enable in enable_dict.items():
+        cmd = subst_opt(cmd, name, enable, opts_dict.get(name))
+
+    return subst_env_vars(cmd, env_vars).replace('\n', ' ')
+
+
+def subst_vars(string, var_dict):
+    '''Apply substitutions in var_dict to string
+
+    If var_dict[K] = V, then <K> will be replaced with V in string.'''
+    for key, value in var_dict.items():
+        string = string.replace('<{}>'.format(key), value)
+    return string
+
+
+def get_yaml_for_simulator(simulator, yaml_path):
+    '''Read yaml at yaml_path and find entry for simulator'''
+    logging.info("Processing simulator setup file : %s" % yaml_path)
+    for entry in read_yaml(yaml_path):
+        if entry.get('tool') == simulator:
+            return entry
+
+    raise RuntimeError("Cannot find RTL simulator {}".format(simulator))
+
+
+def get_simulator_cmd(simulator, yaml_path, enables):
+    '''Get compile and run commands for the testbench
+
+    simulator is the name of the simulator to use. yaml_path is the path to a
+    yaml file describing various command line options. enables is a dictionary
+    keyed by option names with boolean values: true if the option is enabled.
+
+    Returns (compile_cmds, sim_cmd), which are the simulator commands to
+    compile and run the testbench, respectively. compile_cmd is a list of
+    strings (multiple commands); sim_cmd is a single string.
+
+    '''
+    entry = get_yaml_for_simulator(simulator, yaml_path)
+    env_vars = entry.get('env_var', '')
+
+    return ([subst_cmd(arg, enables, entry['compile'], env_vars)
+             for arg in entry['compile']['cmd']],
+            subst_cmd(entry['sim']['cmd'], enables, entry['sim'], env_vars))
+
+
+def rtl_compile(compile_cmds, output_dir, lsf_cmd, opts):
+    """Compile the testbench RTL
+
+    Args:
+      compile_cmd : Compile commands
+      output_dir  : Output directory of the ELF files
+      lsf_cmd     : LSF command to run compilation
+      opts        : Compile options for the generator
+    """
+    # Compile the TB
+    logging.info("Compiling TB")
+    for cmd in compile_cmds:
+        cmd = subst_vars(cmd,
+                         {
+                             'out': output_dir,
+                             'cmp_opts': opts
+                         })
+        logging.debug("Compile command: %s" % cmd)
+        run_cmd(cmd)
```

Comment:
Should `lsf_cmd`be prepended to the compile command?

Reply:
That happens at the bottom of get_test_sim_cmd

Reply:
Ah, good point. Although, in my defence, this is true about the version on master too! :-) I'll fix that up and add a last patch to the set.